neps/nep-0491.md

```diff
@@ -0,0 +1,359 @@
+
+Instead of deploying contracts on the user account, one could build a similar
+solution that uses zero balance accounts and a single master contract that
+performs all smart contract functionality required. This master contract can
+implement the [Storage Management]
+(https://nomicon.io/Standards/StorageManagement) standard to limit storage usage
+per user.
+
+This solution is not as flexible. The master account cannot make cross-contract
+function calls with the user id as the predecessor.
+
+### Alternative: Move away from storage staking
+
+We could also abandon the concept of storage staking entirely. However, coming
+up with a scalable, sustainable solution that does not suffer from the same
+refund problems is hard.
+
+One proposed design is a combination of zero balance accounts and code sharing
+between contracts. Basically, if somehow the deployed code is stored in a way
+that does not require storage staking by the user themself, maybe the per-user
+state is small enough to fit in the 770 bytes limit of zero balance accounts.
+(Questionable for non-trivial use cases.)
+
+This alternative is much harder to design and implement. The proposal that has
+gotten the furthest so far is [Ephemeral
+Storage](https://github.com/near/NEPs/pull/485), which is pretty complicated and
+does not have community consensus yet. Nobody is currently working on moving it
+forward. While we could wait for that to eventually make progress, in the
+meantime, the community is held back in their innovation because of the refund
+problem.
+
+## Future possibilities
+
+- We might want to add the possibility to make non-refundable balance transfers
+  from within a smart contract. This would require changes to the WASM smart
+  contract to host interface. Since removing anything from there is virtually
+  impossible, we shouldn't be too eager in adding it there but if there is
+  demand for it, we certainly can do it without much trouble.
+- We could later add the possibility to refund the non-refundable tokens to the
+  account who sent the tokens initially.
+- We could allow sending non-refundable balance to existing accounts.
+- If (cheap) code sharing between contracts is implemented in the future, this
+  proposal will most likely work well in combination with that. Per-user data
+  will still need to be paid for by the user, which could be sponsored as
+  non-refundable balance without running into refund abuse.
+
+
+## Consequences
+
+
+### Positive
+
+- Businesses can sponsor new user accounts without the user being able to steal
+  any tokens.
+
+### Neutral
+
+- Non-refundable tokens are removed from the circulating supply, i.e. burnt.
+
+### Negative
+
+- Understanding a user's balance become even more complicated than it already
+  is. (Instead of only `amount` and `locked`, there will be a third component.)
+- There is no incentive anymore to delete an account and its state when the
+  backing tokens are not refundable.
```

Comment:
I think the negative aspects of:

> - Understanding a user's balance become even more complicated than it already
>   is. (Instead of only `amount` and `locked`, there will be a third component.)

is the users experience, since the user must now know the nature of another balance in their account and have and understanding of what is possible with it and what is not.
and for the second statement:

> - There is no incentive anymore to delete an account and its state when the
>   backing tokens are not refundable.
the negative side is that we can have a lot of accounts with their state and no reason for their owners to delete them, since they get nothing out of it, which leaves "dead" accounts in circulation.